service/src/main/java/io/techery/janet/AsyncActionService.java

```diff
@@ -108,23 +116,36 @@ public AsyncActionService(String url, AsyncClient client, Converter converter) {
     @Override protected <A> void cancel(ActionHolder<A> holder) {
         AsyncActionWrapper wrapper = getAsyncActionWrapper(holder);
         runningActions.remove(holder.action());
-        if (wrapper.getResponseEvent() != null) {
+        if (wrapper.hasResponse()) {
             synchronizer.remove(wrapper);
         }
     }
 
     private void sendAction(AsyncActionWrapper wrapper) throws AsyncServiceException, CancelException {
-        String responseEvent = wrapper.getResponseEvent();
-        if (responseEvent != null) {
-            synchronizer.put(responseEvent, wrapper);
-        }
         try {
-            ActionBody actionBody = wrapper.getPayload(converter);
-            byte[] content = actionBody.getContent();
+            ActionBody payloadBody;
+            if (wrapper.isBytesPayload()) {
+                payloadBody = (ActionBody) wrapper.getPayload();
+            } else {
+                payloadBody = converter.toBody(wrapper.getPayload());
+            }
+            Message message;
+            byte[] content = payloadBody.getContent();
             if (wrapper.isBytesPayload()) {
-                client.send(wrapper.getEvent(), content);
+                message = protocol.binaryMessageRule().createMessage(wrapper.getEvent(), content);
             } else {
-                client.send(wrapper.getEvent(), new String(content));
+                message = protocol.textMessageRule().createMessage(wrapper.getEvent(), new String(content));
+            }
+            wrapper.setMessage(message);
+            client.send(message);
+            if (wrapper.hasResponse()) {
+                if (protocol.responseMatcher() == null) {
```

Comment:
can we check this in constructor? Go through roster, e.g. While this is more sane, 100 events iteration is not so good for constructor performance.